midp/gfx.js

```diff
@@ -112,8 +112,35 @@ var currentlyFocusedTextEditor;
         return 1;
     };
 
+    var hasNewFrame = true;
+    var ctxToExecuteOnNewFrame;
+    function gotNewFrame(timestamp) {
+        if (ctxToExecuteOnNewFrame) {
+            var ctx = ctxToExecuteOnNewFrame;
+            ctxToExecuteOnNewFrame = null;
+            window.requestAnimationFrame(gotNewFrame);
+            ctx.execute();
+        } else {
+            hasNewFrame = true;
+        }
+    }
 
-    addUnimplementedNative("com/sun/midp/lcdui/DisplayDevice.refresh0.(IIIIII)V");
+    Native["com/sun/midp/lcdui/DisplayDevice.refresh0.(IIIIII)V"] = function() {
+        // This is a no-op: The foreground display gets drawn directly
+        // to the screen.
+        //
+        // However, we sometimes "draw" more frequently than the
+        // screen can update. By calling `requestAnimationFrame` here,
+        // we ensure that whatever we just drew will stick around for
+        // one frame
+        if (hasNewFrame) {
+            hasNewFrame = false;
+            window.requestAnimationFrame(gotNewFrame);
+        } else {
+            ctxToExecuteOnNewFrame = $.ctx;
+            $.pause(asyncImplStringAsync);
```

Comment:
My understanding is that this could potentially cause us to wait when we don't need to.

Let's say that our screen can refresh once every 20ms, and that all of our application logic means that it takes 21ms from one call of `waitForAnimationFrame` to the next. In this scenario, by the time `waitForAnimationFrame` is called, we can already immediately draw to the screen but we will instead call `window.requestAnimationFrame` and wait for a future opportunity to draw.

This is what `hasNewFrame` and `gotNewFrame` are intended to avoid. The idea is that, if we call `waitForAnimationFrame` but we know that we don't need to wait for a new frame, we should just go ahead and draw without waiting.

Let me know if my analysis is flawed.